plugin/pkg/shoot/resourcereservation/admission.go

```diff
@@ -0,0 +1,295 @@
+	readyFunc          admission.ReadyFunc
+
+	typeDependentReservations bool
+}
+
+var (
+	_ = admissioninitializer.WantsCoreInformerFactory(&ResourceReservation{})
+
+	readyFuncs []admission.ReadyFunc
+)
+
+// New creates a new ResourceReservation admission plugin.
+func New(typeDependentReservations bool) admission.MutationInterface {
+	return &ResourceReservation{
+		Handler:                   admission.NewHandler(admission.Create, admission.Update),
+		typeDependentReservations: typeDependentReservations,
+	}
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (c *ResourceReservation) AssignReadyFunc(f admission.ReadyFunc) {
+	c.readyFunc = f
+	c.SetReadyFunc(f)
+}
+
+// SetCoreInformerFactory gets Lister from SharedInformerFactory.
+func (c *ResourceReservation) SetCoreInformerFactory(f gardencoreinformers.SharedInformerFactory) {
+	cloudProfileInformer := f.Core().V1beta1().CloudProfiles()
+	c.cloudProfileLister = cloudProfileInformer.Lister()
+
+	readyFuncs = append(readyFuncs, cloudProfileInformer.Informer().HasSynced)
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (c *ResourceReservation) ValidateInitialization() error {
+	if c.cloudProfileLister == nil {
+		return errors.New("missing cloudProfile lister")
+	}
+	return nil
+}
+
+// Admit injects default resource reservations into worker pools of shoot objects
+func (c *ResourceReservation) Admit(_ context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if c.readyFunc == nil {
+		c.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !c.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	switch {
+	case a.GetKind().GroupKind() != core.Kind("Shoot"),
+		a.GetSubresource() != "":
+		return nil
+	}
+
+	shoot, ok := a.GetObject().(*core.Shoot)
+	if !ok {
+		return apierrors.NewInternalError(errors.New("could not convert resource into Shoot object"))
+	}
+
+	// Pass if the shoot is intended to get deleted
+	if shoot.DeletionTimestamp != nil {
+		return nil
+	}
+
+	if !c.typeDependentReservations {
+		setStaticResourceReservationDefaults(shoot)
+		return nil
+	}
+
+	if shoot.Spec.Kubernetes.Kubelet != nil && shoot.Spec.Kubernetes.Kubelet.KubeReserved != nil {
+		// Inject static defaults for shoots with shoot-level resource reservations
+		setStaticResourceReservationDefaults(shoot)
+		return nil
+	}
+
+	cloudProfile, err := c.cloudProfileLister.Get(shoot.Spec.CloudProfileName)
+	if err != nil {
+		return apierrors.NewInternalError(fmt.Errorf("could not find referenced cloud profile: %+v", err.Error()))
+	}
+
+	allErrs := field.ErrorList{}
+	workersPath := field.NewPath("spec", "provider", "workers")
+
+	for i := 0; i < len(shoot.Spec.Provider.Workers); i++ {
```

Comment:
```suggestion
	for i, worker := range shoot.Spec.Provider.Workers {
```

Reply:
That does not work as `Workers` has type `[]Worker` such that `worker` would be a copy and not a reference to the original value in the slice. Then `injectResourceReservations` would only modify that copy.